### PR TITLE
CircularBuffer: add resize! method

### PR DESCRIPTION
**Pull request by sairus7**
It is adopted from `Base.resize!`:

    resize!(a::Vector, n::Integer) -> Vector

  Resize `a` to contain `n` elements. If `n` is smaller than the current collection length, the first `n` elements will be retained. If `n` is larger, the new elements are not guaranteed to be initialized.

---
With the following meaning:

    resize!(cb, n)

Resize `CircularBuffer` to the maximum capacity of `n` elements. If `n` is smaller than the current buffer length, the first n elements will be retained.

---
Note that it's not changing the current length of the buffer, but its maximum capacity (length of the underlying Array).

### CODE REVIEW

**Comment by sairus7** [comment]
But maybe we should retain last elements, dropping first ones in a similar manner, as `push!` to a full buffer overrides first elements?

**Comment by oxinabox** [comment]
Can you rebase this off current master?

File: src/circular_buffer.jl [hunk 1]
```diff
@@ -216,3 +216,23 @@ Base.@propagate_inbounds function Base.last(cb::CircularBuffer)
     end
     cb.buffer[_buffer_index(cb, cb.length)]
 end
+"""
```
Comment (oxinabox) [inline src/circular_buffer.jl]:
```suggestion

"""
```

Line breaking

File: src/circular_buffer.jl [hunk 2]
```diff
@@ -216,3 +216,23 @@ Base.@propagate_inbounds function Base.last(cb::CircularBuffer)
...
+    resize!(cb, n)
+
+Resize CircularBuffer to the maximum capacity of n elements. If n is smaller than the current buffer length, the first n elements will be retained.
+"""
+function Base.resize!(cb::CircularBuffer{T}, n::Int) where {T}
+    if n != capacity(cb)
+        buf_new = Vector{T}(undef, n)
+        len_new = min(length(cb), n)
+        for i = 1 : len_new
```
Comment (oxinabox) [inline src/circular_buffer.jl]:
```suggestion
        for i in 1 : len_new
```

File: src/circular_buffer.jl [hunk 3]
```diff
@@ -216,3 +216,23 @@ Base.@propagate_inbounds function Base.last(cb::CircularBuffer)
...
+            @inbounds buf_new[i] = cb[i]
+        end
+
+        cb.buffer = buf_new
+        cb.first = 1
+        cb.capacity = n
+        cb.length = len_new
+    end
+    cb
```
Comment (oxinabox) [inline src/circular_buffer.jl]:
```suggestion
    return cb
```

**Comment by oxinabox** [comment]
Some minor stylistic things.

I think for now this is fine.
If it becomes a performance bottleneck for anyone we can look into improving it then

**Comment by oxinabox** [comment]
Can you also bumpp the patch version number so I can tag a release?

**Comment by sairus7** [comment]
> Can you rebase this off current master?

To what branch should I rebase this commit? Or should I change my fork branch? 
Looks like I can change only the target branch:
image

**Comment by oxinabox** [comment]
Seems like no conflict , so no need to rebase.
(You should add a remote for JuliaCollections, and then rebase off that remote's master)

**Comment by sairus7** [comment]
> Now you have done a merge which makes history ugly, but its not the end of the world,
> so you don't need to rebase now.

Will it make ugly history in the `JuliaCollections:master`?
Sorry, I don't quite understand yet the correct workflow. You can point me to, if there is one.

**Comment by oxinabox** [comment]
> Will it make ugly history in the JuliaCollections:master?

Yeah, but it isn't that big a deal.

> Sorry, I don't quite understand yet the correct workflow. You can point me to, if there is one.

Try:
 - https://www.digitalocean.com/community/tutorials/how-to-create-a-pull-request-on-github
 - https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request

**Comment by eulerkochy** [comment]
@sairus7 still up on this PR?

**Comment by rfourquet** [comment]
Calling this functionality `resize!` seems misleading to me, as the lenght is left unchanged (so the behavior diverges from `resize!` on a vector). What about making this a method of `sizehint!` ?

**Comment by sairus7** [comment]
Sorry, forgot about it, closing since it's merged.

**Pull request closed without merging**